### PR TITLE
Update netty and iogrpc version

### PR DESCRIPTION
**Pull request by dasormeter**
grpc
`io.grpc:grpc-xxxx` to `1.29.0` from (1.22.0)

netty
`io.netty:netty-xxxx` update to to `4.1.50.Final` (from 4.1.37.Final)
`io.netty.netty-tcp-native-boringssl-static` to `2.0.30.Final` (from 2.0.25.Final)

spray
`io.spray:spray-json` to `1.3.5` (from 1.3.3)
`org.sangria-graphql:sangria-spray-json` to `1.0.2` (from 1.0.1)
`org.gnieh:diffson-spray-json` to `3.1.1` (from 3.1.0)

To resolve open vulnerabilities with these versions
### netty-4.1.37.Final vulnerabilities
-  BDSA-2018-4022 (Medium)
-  BDSA-2019-2610 (Medium)
-  BDSA-2019-3119 (CVE-2019-16869) (Medium)
-  BDSA-2020-0130 (Medium)
-  BDSA BDSA-2019-4230 (CVE-2019-20445) (Low)
-  BDSA BDSA-2019-4231 (CVE-2019-20444) (Low)
-  BDSA BDSA-2020-0666 (CVE-2020-11612) (Low)
-  BDSA BDSA-2019-2642 (Low)
-  BDSA BDSA-2019-2649 (Low)
-  BDSA BDSA-2019-2643 (Low)

### CODE REVIEW

**Comment by dasormeter** [comment]
> Test failures look legitimate to me

yup everything looks to be working except that reflection IT : not immediately clear on the exception -- need to investigate further

**Comment by SamirTalwar-DA** [comment]
Seems to be Java dependency hell. Glad you figured out a solution, but I'm worried that we can't seem to upgrade GRPC. 🙁

**Comment by gerolf-da** [comment]
We need to keep `grpc-netty`, `netty-handler` and `netty-tcnative-boringssl-static` in line according to https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty. At least I would prefer to use a combination of versions that is known to work.

**Comment by gerolf-da** [comment]
@dasormeter: I took the liberty to push a commit that explains this inline in `bazel-java-deps.bzl` and `deps.bzl`.

**Comment by gerolf-da** [comment]
so yeah, the error stems from a mismatch between which `protobuf-java` version is expected by `grpc-protobuf` and the that is actually on the classpath. Bumping `protobuf` in `deps.bzl` and `protobuf-java` in `bazel-java-deps.bzl` then results in another error that I'm unable to diagnose, let alone fix: 
```
ERROR: /private/var/tmp/_bazel_gerolf/5bbe2d2271b0b313c8d04d49c3350173/external/local_config_cc_toolchains/BUILD.bazel:18:1: in toolchain rule @local_config_cc_toolchains//:cc-toolchain-armeabi-v7a: alias '@rules_cc//cc:toolchain_type' referring to target '@bazel_tools//tools/cpp:toolchain_type' is not visible from target '@local_config_cc_toolchains//:cc-toolchain-armeabi-v7a'. Check the visibility declaration of the former target if you think the dependency is legitimate
```

@aherrmann-da: do you have any idea what this error means and how easy it is to fix?

**Comment by aherrmann-da** [comment]
> Bumping `protobuf` in `deps.bzl` and `protobuf-java` in `bazel-java-deps.bzl` then results in another error that I'm unable to diagnose, let alone fix:
> 
> ```
> ERROR: /private/var/tmp/_bazel_gerolf/5bbe2d2271b0b313c8d04d49c3350173/external/local_config_cc_toolchains/BUILD.bazel:18:1: in toolchain rule @local_config_cc_toolchains//:cc-toolchain-armeabi-v7a: alias '@rules_cc//cc:toolchain_type' referring to target '@bazel_tools//tools/cpp:toolchain_type' is not visible from target '@local_config_cc_toolchains//:cc-toolchain-armeabi-v7a'. Check the visibility declaration of the former target if you think the dependency is legitimate
> ```
> 
> @aherrmann-da: do you have any idea what this error means and how easy it is to fix?

This looks like it ends up with a version of `rules_cc` that's missing public visibility. Looking at the history of `rules_cc` that target was not always publicly visible:
https://github.com/bazelbuild/rules_cc/blame/cd7e8a690caf526e0634e3ca55b10308ee23182d/cc/BUILD#L15

Could you point me to the code change that triggers that issue? Maybe it causes an older version of `rules_cc` to be pulled in.

**Comment by gerolf-da** [comment]
@aherrmann-da: I bumped the `com_google_protobuf` version from 3.8.0 to 3.11.4 in `deps.bzl`.

**Comment by aherrmann-da** [comment]
@gerolf-da Thanks, turns out that version of protobuf [depends on](https://github.com/protocolbuffers/protobuf/blob/39d730dd96c81196893734ee1e075c34567e59ae/protobuf_deps.bzl#L35-L41) an older version of `rules_cc`, where the `toolchain_type` is indeed [not public](https://github.com/bazelbuild/rules_cc/blob/b7fe9697c0c76ab2fd431a891dbb9a6a32ed7c3e/cc/BUILD#L3-L6), yet. An easy way to fix this is to load rules_haskell dependencies first, those pick a more recent version of rules_cc. I.e. apply the following patch:
``` patch
diff --git a/WORKSPACE b/WORKSPACE
index 53fa3d84362..3149c6cf46a 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -18,14 +18,14 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

 daml_deps()

-load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
-
-protobuf_deps()
-
 load("@rules_haskell//haskell:repositories.bzl", "rules_haskell_dependencies")

 rules_haskell_dependencies()

+load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
+
+protobuf_deps()
+
 register_toolchains(
     "//:c2hs-toolchain",
 )
```
I've also proposed an update upstream, so hopefully the issue will go away in future versions: https://github.com/protocolbuffers/protobuf/pull/7552.

**Comment by dasormeter** [comment]
hmm okay we are back to the failing ReflectionIT problem then @gerolf-da  . 
 You think this is still a classpath stomping issue?

//ledger/sandbox:sandbox-tests_test_suite_src_test_suite_scala_com_digitalasset_platform_sandbox_services_reflection_ReflectionIT.scala FAILED in 15.6s

**Comment by dasormeter** [comment]
> Method threw 'java.lang.NoClassDefFoundError' exception. Cannot evaluate io.grpc.reflection.v1alpha.ServerReflectionRequest.toString()

when trying to build ServerReflectionRequest

> ServerReflectionRequest.newBuilder().setHost("127.0.0.1").setListServices("").build()

**Comment by gerolf-da** [comment]
Yes, it means that there is still a `3.8.0` artifact on the classpath it seems.

**Comment by gerolf-da** [comment]
No wait, I haven't pushed the update for `com_google_protobuf` yet, because it keeps on failing. Let me push that.

**Comment by dasormeter** [comment]
> No wait, I haven't pushed the update for `com_google_protobuf` yet, because it keeps on failing. Let me push that.

looks like compilation worked and tests pass on windows and mac osx.  bazel formatting error was causing linux to stop early -- just pushed formatting fix for that

**Comment by dasormeter** [comment]
> > No wait, I haven't pushed the update for `com_google_protobuf` yet, because it keeps on failing. Let me push that.
> 
> looks like compilation worked and tests pass on windows and mac osx. bazel formatting error was causing linux to stop early -- just pushed formatting fix for that

woohoo!  all passed now

File: WORKSPACE [hunk 1]
```diff
@@ -18,14 +18,14 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
...
-protobuf_deps()
-
 load("@rules_haskell//haskell:repositories.bzl", "rules_haskell_dependencies")
 
 rules_haskell_dependencies()
 
+load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
+
+protobuf_deps()
+
```
Comment (stefanobaghino-da) [inline WORKSPACE:28]: Did you move it for a specific reason?

File: bazel-java-deps.bzl [hunk 2]
```diff
@@ -72,20 +66,48 @@ def install_java_deps():
...
+            #
+            # 2. To keep TLS for the Ledger API Server working, the following three artifacts need be updated
+            # in sync according to https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty
+            #
+            # * io.grpc:grpc-netty
+            # * io.netty:netty-handler
+            # * io.netty:netty-tcnative-boringssl-static
+            #
+            # This effectively means all io.grpc:*, io.netty:*, and `com.google.protobuf:protobuf-java
+            # need to be updated with careful consideration.
```
Comment (stefanobaghino-da) [inline bazel-java-deps.bzl:85]: Thanks!

File: deps.bzl [hunk 3]
```diff
@@ -155,11 +155,12 @@ def daml_deps():
...
-            sha256 = "1e622ce4b84b88b6d2cdf1db38d1a634fe2392d74f0b7b74ff98f3a51838ee53",
-            strip_prefix = "protobuf-3.8.0",
-            urls = ["https://github.com/google/protobuf/archive/v3.8.0.zip"],
+            sha256 = "60d2012e3922e429294d3a4ac31f336016514a91e5a63fd33f35743ccfe1bd7d",
+            # changing this version needs to be in sync with protobuf-java and grpc dependencies in bazel-java-deps.bzl
+            strip_prefix = "protobuf-3.11.0",
+            urls = ["https://github.com/google/protobuf/archive/v3.11.0.zip"],
             patches = [
-                "@com_github_digital_asset_daml//bazel_tools:proto-zlib-url.patch",
+                #                "@com_github_digital_asset_daml//bazel_tools:proto-zlib-url.patch",
```
Comment (stefanobaghino-da) [inline deps.bzl:163]: Can you simply remove this please (and the referenced file as well)?

File: WORKSPACE (hunk 1 above)
Comment (dasormeter) [reply WORKSPACE:28]: yeah it is a bazel-fu workaround as advised by @aherrmann https://github.com/digital-asset/daml/pull/6063#issuecomment-633878929

File: bazel-java-deps.bzl (hunk 2 above)
Comment (dasormeter) [reply bazel-java-deps.bzl:85]: this was all herr seitz

File: deps.bzl (hunk 3 above)
Comment (dasormeter) [reply deps.bzl:163]: yeah this was an @aherrmann patch : do we need for any reason or okay to remove this completely now andreas if the PR is working?

Comment (aherrmann-da) [reply deps.bzl:163]: It's okay to remove completely, the comment and the referenced file.

**Pull request merged**